File: backend/src/main/java/com/trind/searchspace/backend/model/filter/Filter.java
package com.trind.searchspace.backend.model.filter;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.trind.searchspace.backend.model.query.targettype.QueryTargetType;

/**
 * Created by Joachim on 2014-09-15.
 */
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(value = TermFilter.class, name = "TERMFILTER"),
        @JsonSubTypes.Type(value = ExistsFilter.class, name = "EXISTSFILTER"),
/*        @JsonSubTypes.Type(value = StatQuery.class, name = "STAT"),
        @JsonSubTypes.Type(value = HistogramQuery.class, name = "HISTOGRAM"),
        @JsonSubTypes.Type(value = ListQuery.class, name = "LIST")*/
})
@JsonIgnoreProperties(ignoreUnknown = true)
public interface Filter {

    Occur getOccur();

    QueryTargetType getSourceTargetType();

    String getField();

    void setSourceTargetType(QueryTargetType sourceTargetType);

    void setField(String field);

}

File: web/src/main/resources/public/application/js/filter.js
/** Filters**/
var FilterTypes = {
    TERMFILTER: "TERMFILTER",
    EXISTSFILTER: "EXISTSFILTER"
};

addFilterType(FilterTypes.EXISTSFILTER, ExistsFilter);
addFilterType(FilterTypes.TERMFILTER, TermFilter);

function Filter(type) {
    var self = this;
    self.type = type;
    self.sourceTargetType;
    self.occur;
}

function TermFilter() {
    var self = this;
    ko.utils.extend(self, new Filter(FilterTypes.TERMFILTER));

    self.field;
    self.value;
}

function ExistsFilter() {
    var self = this;
    ko.utils.extend(self, new Filter(FilterTypes.EXISTSFILTER));

    self.field;
}

/** End filters.*/